File: src/mvc/python/reorderable_treeview/reorderable_tree_view.py
from PyQt6.QtCore import Qt, QModelIndex, QPoint
from PyQt6.QtWidgets import QAbstractItemView, QMenu, QTreeView


class ReorderableTreeView(QTreeView):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setEditTriggers(QAbstractItemView.EditTrigger.NoEditTriggers)

        self.setDragEnabled(True)
        self.setAcceptDrops(True)
        self.setDropIndicatorShown(True)
        self.setDragDropMode(QAbstractItemView.DragDropMode.InternalMove)
        self.setSelectionMode(QAbstractItemView.SelectionMode.SingleSelection)

        self._cut_index = QModelIndex()
        self._drag_index = QModelIndex()

    def move_to_destination(self, source_index, destination_index):
        self.model().move_to_destination(source_index, destination_index)
        self.expandAll()
        self.clearSelection()

    def startDrag(self, event):
        self._drag_index = self.currentIndex()
        super().startDrag(event)

    def dropEvent(self, event):
        position = QPoint(int(event.position().x()), int(event.position().y()))
        index = self.indexAt(position)
        self.move_to_destination(self._drag_index, index)

    def contextMenuEvent(self, event):
        global_position = event.globalPos()
        position = self.viewport().mapFromGlobal(global_position)
        index = self.indexAt(position)

        menu = QMenu(self)
        cut_action = menu.addAction("Cut")
        copy_action = menu.addAction("Copy")
        paste_action = menu.addAction("Paste")
        delete_action = menu.addAction("Delete")

        action = menu.exec(global_position)
        if action == cut_action:
            self._cut_index = index
        elif action == copy_action:
            pass
        elif action == paste_action:
            self.move_to_destination(self._cut_index, index)
        elif action == delete_action:
            self.model().removeRow(index.row(), index.parent())

File: src/mvc/cpp/reorderable_treeview/reorderable_tree_view.cpp
#include "reorderable_tree_view.h"
#include <QDebug>
#include <QMouseEvent>

ReorderableTreeModel::ReorderableTreeModel(const QStringList &headers,
                                       QObject *parent)
    : TreeModel(headers, parent) {}

void ReorderableTreeModel::appendToDestination(QModelIndex sourceIndex,
                                             QModelIndex destinationIndex) {
  auto sourceItem = static_cast<TreeItem *>(sourceIndex.internalPointer());
  auto destinationItem =
      static_cast<TreeItem *>(destinationIndex.internalPointer());
  destinationItem->appendChild(sourceItem);
  emit dataChanged(sourceIndex, sourceIndex);
  emit dataChanged(destinationIndex, destinationIndex);
}

Qt::DropActions ReorderableTreeModel::supportedDropActions() const {
  return Qt::CopyAction | Qt::MoveAction;
}

Qt::ItemFlags ReorderableTreeModel::flags(const QModelIndex &index) const {

  Qt::ItemFlags defaultFlags = QAbstractItemModel::flags(index);

  if (index.isValid())
    return Qt::ItemIsDragEnabled | Qt::ItemIsDropEnabled | defaultFlags;
  else
    return Qt::ItemIsDropEnabled | defaultFlags;
}

ReorderableTreeView::ReorderableTreeView(QWidget *parent) : QTreeView(parent) {

  setDragEnabled(true);
  setAcceptDrops(true);
  setDropIndicatorShown(true);
  setDragDropMode(QAbstractItemView::InternalMove);
}

ReorderableTreeView::~ReorderableTreeView() {}

ReorderableTreeModel *ReorderableTreeView::model() const {
  return static_cast<ReorderableTreeModel *>(QTreeView::model());
}

void ReorderableTreeView::setModel(ReorderableTreeModel *model) {
  QTreeView::setModel(model);
}

void ReorderableTreeView::startDrag(Qt::DropActions supportedActions) {
  dragIndex = currentIndex();
  QTreeView::startDrag(supportedActions);
}

void ReorderableTreeView::dragEnterEvent(QDragEnterEvent *event) {
  QTreeView::dragEnterEvent(event);
}

void ReorderableTreeView::dropEvent(QDropEvent *event) {
  dropIndex = indexAt(event->position().toPoint());
  model()->appendToDestination(dragIndex, dropIndex);
  reset();
  expandAll();
}

void ReorderableTreeView::dragMoveEvent(QDragMoveEvent *event) {
  QTreeView::dragMoveEvent(event);
}

void ReorderableTreeView::dragLeaveEvent(QDragLeaveEvent *event) {
  QTreeView::dragLeaveEvent(event);
}
